[Gujar Parth Shailesh] iP

parth-io opened the pull request:
## Duke has more tasks now!

> Welcome to the new age

It now supports these features and more:

- **adding** todos, events or deadlines; ~~deleting~~ tasks; _searching_ for tasks
- ___easier___ to use, with better error messages
- persists data in local storage

Steps to use :bowtie::
1. Download it from [here](https://github.com/parth-io/ip/releases/download/A-Jar/duke.jar)
2. Double-click it or run `$ java -jar duke.jar` in your bash terminal from the folder where the file is.
3. Add your tasks!

Things to improve:
- [x] Code conventions
- [ ] Better use of OOP principles

```java
public void printGreeting() {
        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println("Hello from\n" + logo);
    }
```

Commit:
Add Gradle support

parth-io committed:
Add Level-1
Add to .gitignore

parth-io committed:
Add Level-2

parth-io committed:
Add Level-3

parth-io committed:
Add Level-4

parth-io committed:
Add A-TextUiTesting

parth-io committed:
Add Level-5

parth-io committed:
Add Level-6

parth-io committed:
Add A-Enums

parth-io committed:
Refactor code for task commands to create new tasks
Refactor CustomMessageException to outside of Duke.java

parth-io committed:
Refactor code for task commands and follow coding conventions

Common behaviours in task commands cause code duplication.

Let's simplify how the code parses task commands.
Let's follow coding conventions as well.

parth-io committed:
Refactor Task class and subclasses

The enum functionality is split across the Enum class and the Task classes.

Let's link the two together.

Let's add the ability for tasks to generate a storage string also.

parth-io committed:
Implement the save feature

The code does not persist the tasks.

Let's refactor the Duke class and abstract away the TaskList into a separate class.

Let's also create an InputParser class to read the input from user or from file.

parth-io committed:
Implement the date/time feature

The code does not parse the date/time the user enters.

Let's refactor add LocalDateTime and parse the user input.

parth-io committed:
Merge branch 'branch-Level-7'

parth-io committed:
Merge branch 'branch-Level--8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	text-ui-test/EXPECTED.TXT
#	text-ui-test/input.txt

parth-io committed:
Fix branch-Level-8 merge conflicts

parth-io committed:
Merge branch 'add-gradle-support'

parth-io committed:
Refactor TaskList and Parser into Storage and Ui

TaskList and InputParser are doing too much in one class.

Let's move functionality out to other classes to achieve separation of concerns.

parth-io committed:
Divide classes into packages

parth-io committed:
Add JUnit tests

The code has no tests.

Let's test methods in Parser and TaskList

parth-io committed:
Create JAR file and improve printing

parth-io committed:
Add find functionality

Find is not a recognised command.

Let's add it.

parth-io committed:
Add JavaDocs to non-private classes, methods, and fields

parth-io committed:
Fix minor Java coding style violations

parth-io committed:
Merge branch 'branch-A-CodingStandard'

parth-io committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Parser.java

parth-io committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/TaskList.java

parth-io committed:
Fix wildcard import

riccqi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+                String nextCommand = ui.getNextCommand();
+                ui.showLine();
+                System.out.print(parser.parseUserCommand(nextCommand, parser.breakLoopIndicator));
+                if (parser.breakLoopIndicator.getIsExitCommand()) {
```
```suggestion
                if (parser.breakLoopIndicator.getExitCommand()) {
```
small nit, i think we can drop the `is`

parth-io replied:
Alright, thanks for the comments

riccqi commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+//                Command c = Parser.parse(fullCommand);
+//                c.execute(tasks, ui, storage);
+            } catch (CustomMessageException e) {
+                ui.showError(e.getMessage());
```
could we have taken advantage of the `toString` method in your exception class instead?

riccqi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,167 @@
+        }
+        commands[0] = "";
+        int index;
+        String toPrint;
```
i think would be better to name it a noun, since its a variable

riccqi commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,59 @@
+     * @param message The message to indent
+     * @return A {@code String} representing the indented message
+     */
+    public static String indentedMessage(String message) {
+        return "     " + message;
+    }
```
this is a function, so i would do `indentMessage` instead!

riccqi commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     *
+     */
+    @BeforeEach
```
Good use of the junit annotations!

riccqi reviewed (commented):
LGTM, some small nits to fix!

parth-io committed:
Add GUI

parth-io committed:
Merge branch 'branch-Level-10'

parth-io committed:
Clean up tutorial code

parth-io committed:
Add Checkstyle

The project does not follow the recommended style.

Let's add Checkstyle

parth-io committed:
Merge branch 'branch-A-Checkstyle'

parth-io committed:
GUI: transition to a GUI

Using the CLI-mode is difficult and not visually appealing.

Responses in the CLI mode are also cumbersome to add, with delimitations made using lines and indentations.

Let's use JavaFX to add a GUI. Let's also simplify the generation of Duke's responses in the code.

Using JavaFX is recommended because JavaFX is a versatile and popular library to create a desktop GUI. The resulting UI is simple and minimalistic.

parth-io committed:
The code has no assertions.

Assertions are a useful way of checking the code is executing as per our expectations.

Let's add assertions to TaskList and Parser.java.

These files have complicated logic.

parth-io committed:
The code has some quality issues.

We need to improve readability.

Let's remove dead code and unnecessary comments. Let's also avoid complicated logical expressions. Let's also make the happy path more prominent, where applicable.

This commit also fixes #1

parth-io committed:
Merge pull request #2 from parth-io/branch-A-Assertions

The code has no assertions.

parth-io committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Storage.java

parth-io committed:
Merge pull request #3 from parth-io/branch-A-CodeQuality

The code has some quality issues.

parth-io committed:
Create gradle.yml

Let's set up GitHub Actions with the recommended template file.

parth-io committed:
The code checks for invalid number of arguments in multiple places.

Let's add this check to Command.java.

parth-io committed:
Parser.java has a very long method.

This is poor code quality.

Let's refactor the different Command executions to a separate package.

parth-io committed:
Let's polish the use of Streams in TaskList.

Let's also refactor the different TaskList related classes and methods to a separate package.

parth-io committed:
Merge pull request #4 from parth-io/branch-A-Streams

Let's improve the use of Streams in TaskList.

parth-io committed:
Let's fix the GitHub yml file.

parth-io committed:
Currently, the code has major issues with code extensibility and flexibility. The code for TaskList and Storage is tightly coupled. The Parser class and TaskList class were doing too much.

This prevents the code from being easily modified.

Let's simplify TaskList and Parser. Let's also add to the execute method of the Command classes, so only the execute method dictates the flow.

This approach allows the code to adhere more closely to the SLAP principle.

parth-io committed:
Let's improve the code structure by creating a new package duke.gui

parth-io committed:
The user is currently unable to undo the previous command.

Let's add an UNDO command.

Running an undoable command will store the previous TaskList to a backup file. Storing the previous state over creating an inverse for each command is preferable, as it is simpler to implement.

parth-io committed:
Add documentation to the project.

parth-io committed:
Let's add more tests and refactor Storage.java

parth-io committed:
Let's add Win and Mac OpenJFX libraries

parth-io committed:
Let's improve the test files

parth-io committed:
Merge branch 'temp'

parth-io committed:
Let's fix the Windows test issue

parth-io committed:
Let's fix the Windows test issue
